File: hw/vizcaino/u3/hw21FASPyJs/hw21JavaScript/model/farmAnimal.js
export class FarmAnimal {
    constructor(id, breed, bornOn, weight, slaughterHouse, product, cuts) {
        this.id = id;
        this.breed = breed;
        this.bornOn = bornOn;
        this.weight = weight;
        this.slaughterHouse = slaughterHouse;
        this.product = product;
        this.cuts = cuts;
    }

    getAgeInMonths() {
        if (!this.bornOn) {
            throw new Error("bornOn is not set");
        }

        const today = new Date();

        if (this.bornOn > today) {
            throw new Error("bornOn cannot be in the future");
        }

        return (
            (today.getFullYear() - this.bornOn.getFullYear()) * 12 +
            (today.getMonth() - this.bornOn.getMonth())
        );
    }

    feed(food) {
        throw new Error("Method not implemented");
    }

    toString() {
        return `FarmAnimal{id=${this.id}, breed=${this.breed}, bornOn=${this.bornOn.toDateString()}, weight=${this.weight}}`;
    }
}

File: hw/vizcaino/u3/hw21FASPyJs/hw21Python/model/farm_animal.py
from abc import ABC, abstractmethod
from datetime import date


class FarmAnimal(ABC):
    def __init__(
        self,
        id,
        breed,
        born_on,
        weight,
        slaughter_house,
        product,
        cuts,
    ):
        self.id = id
        self.breed = breed
        self.born_on = born_on
        self.weight = weight
        self.slaughter_house = slaughter_house
        self.product = product
        self.cuts = cuts

    def get_age_in_months(self):
        if self.born_on is None:
            raise ValueError("born_on is not set")

        today = date.today()

        if self.born_on > today:
            raise ValueError("born_on cannot be in the future")

        return (
            (today.year - self.born_on.year) * 12
            + today.month
            - self.born_on.month
        )

    @abstractmethod
    def feed(self, food):
        pass

    def __str__(self):
        return (
            f"FarmAnimal(id={self.id}, breed={self.breed}, "
            f"born_on={self.born_on}, weight={self.weight})"
        )